[useRenderer] Add public hook (#1418)
mui/base-ui: docs/src/app/(public)/(content)/react/utils/use-renderer/page.mdx

# useRenderer

<Subtitle>Utility for adding Base UI like features in a custom build components.</Subtitle>
<Meta
  name="description"
  content="Utility for adding Base UI like features in a custom build components."
/>

> Base UI provides few common features across all components:
> 
> - support for a [callback](/react/handbook/styling#css-classes) on the `className` prop where developers can generate classes based on the state of the components
> - [render](/react/handbook/composition) prop that allows developers to override the default rendered element of the component
> - adds data-attributes for the component's state that can be used as style hooks
> 
> The `useRenderer` hook allows you to add these feature on your custom built components, so that you can provide consistenc experience across your library.

Reviewer: ```suggestion
The `useRenderer` hook allows you to support the same features Base UI provides across all components so that you can also have a consistent experience in your custom components:

- A [render](/react/handbook/composition) prop to override the default rendered element.
- A [callback](/react/handbook/styling#css-classes) on the `className` prop that enables passing dynamically generated CSS classes based on the component's state.
- Adds [data-attributes](https://base-ui.com/react/handbook/styling#data-attributes) that map to the component's state that can be used as style hooks.
```

## API reference

### Input parameters

<PropsReferenceTable
  data={{
    className: {
      type: 'string | ((state: State) => string)',
      description:

>         'The class name to apply to the rendered element. Can be a string or a function that accepts the state and returns a string.',

Reviewer: Should we use the same description used in component prop tables?

```suggestion
        'CSS class applied to the element, or a function that returns a class based on the component’s state.',
```

    },
    render: {
      type: 'RenderProp<State>',

>       description: 'The render prop or React element to override the default element.',

Reviewer: In components' prop tables we say:

```suggestion
      description: 'Allows you to replace the component’s HTML element\nwith a different tag, or compose it with another component.\n\nAccepts a `ReactElement` or a function that returns the element to render.',
```
